src/hooks/core/post/usePostHook.tsx

```diff
@@ -0,0 +1,129 @@
+          feedsContext.feeds[index] = {...feedsContext?.feeds[index], is_following_target: true};
+        }
+      });
+      setFollow(data);
+    } else {
+      const newFeed = {...feedsContext?.feeds[indexFeed], is_following_target: true};
+      feedsContext.feeds[indexFeed] = newFeed;
+      followUserAnon(dataFollowAnon);
+    }
+    setMainFeeds(feedsContext.feeds, feedsContextDispatch);
+  };
+
+  const followUnfollowTopic = async (item: any, topic: string, offset: number) => {
+    const id = removePrefixTopic(topic);
+    const feedTopic = feedsContext.topicFeeds
+      ? feedsContext.topicFeeds.filter((feed) => feed?.topics?.includes(id))
+      : [];
+    const feedMain = feedsContext.feeds;
+
+    const user_id = item?.actor?.id;
+    const username = item?.actor?.data?.username;
+    const data = {
+      user_id_follower: myProfile?.user_id,
+      user_id_followed: user_id,
+      username_follower: myProfile?.username,
+      username_followed: username,
+      follow_source: 'feed'
+    };
+    const dataFollowAnon = {
+      follow_source: 'post',
+      post_id: item?.id
+    };
+    const indexFeed = feedTopic?.findIndex((feed) => {
+      return feed?.id === item?.id;
+    });
+    const feedData = feedTopic[indexFeed];
+    if (feedData?.is_following_target) {
+      if (!feedData?.anon_user_info_color_name) {
+        feedTopic.forEach((feed, index) => {
+          if (feed?.actor?.id === user_id) {
+            feedTopic[index] = {...feedTopic[index], is_following_target: false};
+          }
+        });
+        feedMain.forEach((feed, index) => {
+          if (feed?.actor?.id === user_id) {
+            feedMain[index] = {...feedMain[index], is_following_target: false};
+          }
+        });
+        setUnFollow(data);
+      } else {
+        feedTopic[indexFeed] = {...feedTopic[indexFeed], is_following_target: false};
+        feedMain[indexFeed] = {...feedMain[indexFeed], is_following_target: false};
+        unfollowUserAnon(dataFollowAnon);
+      }
+    } else if (!feedData?.anon_user_info_color_name) {
+      feedTopic.forEach((feed, index) => {
+        if (feed?.actor?.id === user_id) {
+          feedTopic[index] = {...feedTopic[index], is_following_target: true};
+        }
+      });
+      feedMain.forEach((feed, index) => {
+        if (feed?.actor?.id === user_id) {
+          feedMain[index] = {...feedMain[index], is_following_target: true};
+        }
+      });
+      setFollow(data);
+    } else {
+      feedTopic[indexFeed] = {...feedTopic[indexFeed], is_following_target: true};
+      feedMain[indexFeed] = {...feedMain[indexFeed], is_following_target: true};
+      followUserAnon(dataFollowAnon);
+    }
+    TopicPageStorage.set(id?.toLowerCase(), feedTopic, offset);
+    setTopicFeeds(feedTopic, feedsContextDispatch);
+    setMainFeeds(feedMain, feedsContextDispatch);
+  };
+
+  return {
+    followUnfollow,
+    followUnfollowTopic
+  };
+};
+
+export default usePostHook;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `usePostHook` hook seems to be well implemented. However, there are a few areas that could be improved for better readability and maintainability.

1. **Avoid direct mutation of state**: In React, you should avoid directly mutating the state as it can lead to unexpected behavior. Instead, create a new copy of the state, modify it, and then set the state with the new copy. For example, in lines 35, 41, 47, 52, 87, 92, 97, 98, 104, 109, 114, and 115, you're directly modifying `feedsContext.feeds` and `feedTopic`, which are part of the state. 

2. **DRY principle violation**: The `followUnfollow` and `followUnfollowTopic` functions have similar logic. You could extract the common logic into a separate function to adhere to the DRY (Don't Repeat Yourself) principle. This would make your code more maintainable and easier to understand.

Here's an example of how you might refactor the `followUnfollow` function to address these issues:

```diff
  const followUnfollow = async (item: any) => {
    const user_id = item?.actor?.id;
    const username = item?.actor?.data?.username;
    const data = {
      user_id_follower: myProfile?.user_id,
      user_id_followed: user_id,
      username_follower: myProfile?.username,
      username_followed: username,
      follow_source: 'feed'
    };
    const dataFollowAnon = {
      follow_source: 'post',
      post_id: item?.id
    };
    const indexFeed = feedsContext?.feeds?.findIndex((feed) => {
      return feed?.id === item?.id;
    });
    const feedData = feedsContext?.feeds[indexFeed];
    if (feedData?.is_following_target) {
      if (!feedData?.anon_user_info_color_name) {
        const newFeeds = feedsContext?.feeds.map((feed) => {
          if (feed?.actor?.id === user_id) {
            return {...feed, is_following_target: false};
          }
          return feed;
        });
        setUnFollow(data);
        setMainFeeds(newFeeds, feedsContextDispatch);
      } else {
        const newFeed = {...feedsContext?.feeds[indexFeed], is_following_target: false};
        const newFeeds = [...feedsContext.feeds];
        newFeeds[indexFeed] = newFeed;
        unfollowUserAnon(dataFollowAnon);
        setMainFeeds(newFeeds, feedsContextDispatch);
      }
    } else if (!feedData?.anon_user_info_color_name) {
      const newFeeds = feedsContext?.feeds.map((feed) => {
        if (feed?.actor?.id === user_id) {
          return {...feed, is_following_target: true};
        }
        return feed;
      });
      setFollow(data);
      setMainFeeds(newFeeds, feedsContextDispatch);
    } else {
      const newFeed = {...feedsContext?.feeds[indexFeed], is_following_target: true};
      const newFeeds = [...feedsContext.feeds];
      newFeeds[indexFeed] = newFeed;
      followUserAnon(dataFollowAnon);
      setMainFeeds(newFeeds, feedsContextDispatch);
    }
  };
```

This refactoring addresses the direct state mutation issue by creating a new array `newFeeds` and updating it instead of directly modifying `feedsContext.feeds`. It also makes the code more readable by reducing the number of nested if-else statements. Similar changes can be made to the `followUnfollowTopic` function.


<!-- This is an auto-generated comment by OSS CodeRabbit -->